Fix provider config

Tahvohck opened the pull request:
Fixes some issues with ProviderConfig saving files (or not) when they didn't already exist.
Also improves the way it's initialized and adds an optional validation pass whenever saving and loading.
DemoProvider is now the gold standard for Providers and as such has had multiple unit tests added.

Tahvohck committed:
Fix: ProviderConfig now ACTUALLY saves

Tahvohck committed:
heheh funny

Tahvohck committed:
Remove save() changing SHOULD_SAVE

I can't think of any good reason why save() should be changing that. It edits no properties on the object and we probably shouldn't just assume it's safe to turn saving back on once complete.

Tahvohck committed:
Change how save location is passed

This way configs don't need to describe a constructor if they have no special constructor code. It also enforces a save location being passed in in the first place.

Tahvohck committed:
Implement ProviderConfig on demo provider

The demo provider should always be up to date as the gold standard for a provider.

Tahvohck committed:
Move config into you know. The config.

Tahvohck committed:
add extra comments

Tahvohck committed:
Add method for configs to validate (optional)

Tahvohck committed:
Wrap unit tests in import.meta.main block

Tahvohck committed:
Revert "Wrap unit tests in import.meta.main block"

This reverts commit 1fcee0d91bede9e2db07e8b6206d9829b31b1285.
It turns out that during testing, import.meta.main is false.

Tahvohck committed:
Add unit tests

Tahvohck committed:
cherry-pick aa8ad8a1e6687ce4371ec45f48fd7a88cd5528d4

Tahvohck committed:
make validate actually optional instead of weirdly optional

I won the coin flip :)

Mampfinator commented on src/DonationProvider.ts:
```diff
@@ -128,37 +141,42 @@ export class ProviderConfig {
-
             for (const [key, value] of Object.entries(json)) {
                 Reflect.set(config as object, key, value);
             }
```
We should probably validate after setting all loaded properties now that it's available to us, in case people mess with config files and input invalid values.

Tahvohck replied:
Yeah, that's a good point.

Tahvohck replied:
![image](https://github.com/user-attachments/assets/58e58605-9409-4ed2-a949-96b40c800c27)
![image](https://github.com/user-attachments/assets/015c5f03-d904-4483-a681-87f096826c33)
A good point and that's why it's already in there

Mampfinator replied:
I may in fact be blind

Tahvohck committed:
Break config class out into its own file

In preparation for making ProviderConfig into a general locally saved config, there's nothing inherently linked to the providers other than that's where we initially ran into needing them.

Tahvohck committed:
Rename class to reflect the change to generic

Tahvohck committed:
Update comments and documentation

Tahvohck committed:
Fix imports

these worked fine in production but vscode thought they were relative imports

Tahvohck committed:
Skip auto validation during internal plumbing

Tahvohck committed:
Default behavior should be to save unless internally disabled.

Tahvohck committed:
Move tests into dedicated test file

Tahvohck committed:
assertRejects needs to be awaited

Tahvohck committed:
Add message to test error to avoid blank lines

Tahvohck commented:
Don't mind those force pushes, I'm just stupid and forgot to remove unused imports when breaking the code out. Twice.

Tahvohck committed:
default this back to false

After discussion this should only be true if a clean load from disk was successful.

Tahvohck committed:
Rename to be more explicit about behavior

Tahvohck committed:
break code into reusable function

Tahvohck committed:
Remove unused test

Tahvohck committed:
clean tests better at start

Tahvohck committed:
fix broken tests

Tahvohck committed:
Load files without fucking user config

Previously this was very likely to overwrite user configuration. JSON parse has been broken out from file load so that file can safely be created if not present without overwriting the file if the json fails.

Tahvohck committed:
Update test name to be clearer

Tahvohck committed:
Merge branch 'master' into fix-provider-config

Tahvohck committed:
Add complex case test (broken)

Tahvohck committed:
remove ugly as sin typing

honestly I blame it being 2:30am for choosing that typing, but now it's gone

Mampfinator committed:
allow `SavedConfig` to also trap and save on nested property sets
and update the complicated test accordingly. We can't restore classes on load, so that has been removed.

Mampfinator committed:
improve order of some calls/statements to reduce unnecessary work

Mampfinator committed:
guard against potentially indexing null properties

Mampfinator committed:
Revert "improve order of some calls/statements to reduce unnecessary work"

This reverts commit c74dc99d2daa2dc70aaf81a8b16d4383a081c000.

Tahvohck committed:
Merge branch 'eats/nested-property-saving' into fix-provider-config

Closes #8
This should be good enough for most cases. Deep nesting will get rid of most of
the potential fuckery.

Tahvohck committed:
automatic format pass

Tahvohck committed:
mild whitespace and comment cleanup

Tahvohck committed:
BUGFIX: skip validation and saving on symbols
